### ISSUE TITLE
What is the major consideration when choosing Hamming window with no periodic?

### ISSUE THREAD

**Issue by 20050710212** [anicolson/DeepXi#37]
Hi, Aaron, thanks for your fantastic work.
I tried to change the window you used from 
`functools.partial(window_ops.hamming_window, periodic=False)`
to 
`functools.partial(window_ops.hann_window, periodic=True)`
Then, it seems that the convergence becomes slower than before. 

By comparison, the Hamming window does a better job of cancelling the nearest side lobe, but a poorer job of canceling any others. Is this the reason why you choose Hamming?  Thanks.

**Comment by anicolson** [comment]
To be honest, I haven't compared the performance. I would not expect there to be a significant performance difference between either. 

The symmetric Hamming window was used simply because it is commonly used for speech enhancement.

Please see page 11 of the following for a detailed description of both windows. https://download.ni.com/evaluation/pxi/Understanding%20FFTs%20and%20Windowing.pdf

See the following for symmetry:
https://en.wikipedia.org/wiki/Window_function#Symmetry.

Hope this helps,
Aaron.

**Comment by anicolson** [comment]
That is not to say Hann window is not commonly used in speech enhancement.

**Comment by 20050710212** [comment]
Thanks for your reply. I read the same materials as you recommend. 
I accompIished some speech enhancement projects and used the squared root of Hann window with periodic for the analysis and synthesis. 

Can I discuss with you about the **Mapped a priori SNR**?

I read your two papers, _DEEP XI AS A FRONT-END FOR ROBUST AUTOMATIC SPEECH RECOGNITION_ and _Deep learning for minimum mean-square error approaches to speech enhancement_. I can not find detail about the effects of mapped a priori SNR on convergence.

I checked the statistic information and tried to generate the new one according to my training sets. The **mu_hat** are almost the same, but the **sigma_hat** are different for different windows. Furthermore, I tried to generate them many times. The variations of **sigma_hat** while using **Hamming** window is smaller than that while using **Hann** window. I suppose **sigma_hat** can affect the convergence speed significantly. I am not sure about these findings.

Could you share some experience about the **Mapped a priori SNR**? 

Thank you.

**Comment by anicolson** [comment]
Hi, sorry for the late reply.

If one tries to estimate the a priori SNR without mapping, the network will not converge well, i.e. linear output layer with MSE loss and a priori SNR as the target. But, if you squeeze the a priori SNR between 0 and 1, like the IRM, the network will train.
________________________________

**Comment by anicolson** [comment]
So I did a test recently that will be made available on this repository at a later date. I did this a while ago, but couldn't find the results.

The first test is the instantaneous a priori SNR as the training target with mean squared error loss. The loss does not converge and the network does not train.

The first test is the instantaneous a priori SNR in **dB** as the training target with mean squared error loss. The loss does not converge and the network does not train.  

And as you know, using CDF of instantaneous a priori SNR (dB) with cross-entropy works well. Mean squared error could be used for this case as well.

**Comment by 20050710212** [comment]
It is unusual to use cross-entropy in a regression task and it puzzled me for a long time. I plot both of the cross entropy and squared error for different target and estimate values from 0.1 to 0.9 and use 0.1 as a step. An x-axis value represents an estimate values, and each line in figures expresses one target value. 

From the figures, the ideal loss of MSE is zero. However, it is close to 0.47 for cross entropy and depends on the SNR distribution. Besides, 0.5 seems to be a magic number in cross entropy loss.

I am interested to know why you pick cross entropy as the default loss function in DeepXi.

Thank you very much.

![image](https://user-images.githubusercontent.com/4574648/82786590-7cd7f700-9e97-11ea-9d3d-c3452c7b2b3c.png)
![image](https://user-images.githubusercontent.com/4574648/82786619-895c4f80-9e97-11ea-9514-7180ffafadc8.png)

**Comment by yunzqq** [comment]
> It is unusual to use cross-entropy in a regression task and it puzzled me for a long time. I plot both of the cross entropy and squared error for different target and estimate values from 0.1 to 0.9 and use 0.1 as a step. An x-axis value represents an estimate values, and each line in figures expresses one target value.
> 
> From the figures, the ideal loss of MSE is zero. However, it is close to 0.47 for cross entropy and depends on the SNR distribution. Besides, 0.5 seems to be a magic number in cross entropy loss.
> 
> I am interested to know why you pick cross entropy as the default loss function in DeepXi.
> 
> Thank you very much.
> 
> ![image](https://user-images.githubusercontent.com/4574648/82786590-7cd7f700-9e97-11ea-9d3d-c3452c7b2b3c.png)
> ![image](https://user-images.githubusercontent.com/4574648/82786619-895c4f80-9e97-11ea-9514-7180ffafadc8.png)

I am the co-author. I think that you just need to some experiments. If it is OK, you can also choose MSE as loss.   Deep learning technqiue may also need many explainations.  I know that you may concern that cross-entropy is mainly used for two classed task. But the output is also like a possibility between 0-1 to predict 0 (close to 0) or 1 (close to 1) (never be the exact 0 and 1 ). It is easy to explain that we just want to output ranges from 0-1.   But you can use MSE, if you think it is OK.

**Comment by anicolson** [comment]
Originally, the cross-entropy loss function was used to estimate the ideal binary mask (IBM) for computational auditory scene analysis (CASA). This is a binary classification task, i.e. each target can be 1 (speech dominant) or 0 (noise dominant). This was later extended to the ideal ratio mask, which is just a soft version of the IBM, i.e. values in between 1 and 0 indicate a ratio of speech to noise. The cross-entropy loss function was also used. Cross-entropy loss can be thought of as measuring the difference between two probabilities. For the IRM, the higher the value (i.e. the closer the IRM is to 1), the higher the probability of the spectral component being (completely) speech. Cross-entropy can also be thought of as an error measure restricted to input values in the interval [0,1]. For the IRM and IBM cases, cross-entropy was also used with a sigmoid activated output layer. With a DNN that estimates the clean speech magnitude spectrum, a linear/ReLU output layer must be used with MSE. 

For IRM, IBM, and CASA, please see the following: https://ieeexplore.ieee.org/document/6639038.

There is a paper somewhere showing that IRM target + cross-entropy + sigmoid output layer outperforms clean speech magnitude spectrum target + MSE + linear output layer. I will have to find it.

For our case, we are estimating the mapped a priori SNR (i.e. the cumulative distribution function of the instantaneous a priori SNR). This can be thought of similarly to the IRM; a value of 1 indicates completely speech and a value of 0 indicates completely noise. And as IRM target + cross-entropy + sigmoid output layer has been found to outperform clean speech magnitude spectrum target + MSE + linear output layer, I chose to replace the IRM of the first case with the mapped a priori SNR.

Hope this helps,
Aaron.

**Comment by anicolson** [comment]
To add on to this, I have not tested yet which of the following combinations performs best:

a) mapped a priori SNR  target + cross-entropy + sigmoid output.
b) mapped a priori SNR  target + cross-entropy + linear output.
c) mapped a priori SNR  target + MSE + sigmoid output.
d) mapped a priori SNR  target + MSE + linear output.

Maybe someone can test this, if not I can look into this at some point in the future.

**Comment by 20050710212** [comment]
@yunzqq @anicolson Thanks for your detailed reply. It is extremely informative. 

Actually, I tried the **c) mapped a priori SNR target + MSE + sigmoid output**, by merely change loss function as **mean_squared_error** in your project. From my results, the combination **a)** works much better than **c)**. However, the training datasets I used are different from yours.

**Comment by anicolson** [comment]
Hi,

I would like to report an update to this,

I incorrectly forgot to remove the sigmoid layer when training instantaneous a priori SNR + MSE
and instantaneous a priori SNR in dB + MSE.

So after training them correctly, I have found that instantaneous a priori SNR + linear output + MSE does not converge, instantaneous a priori SNR in dB + linear output + MSE does converge, but performs worse than mapped a priori SNR + sigmoid output + Cross-entropy.

Regards,
Aaron.
